src/libos/src/process/elf_file.rs

```diff
@@ -1,95 +1,173 @@
+use scroll::{self, ctx, Pread};
 
-pub use xmas_elf::header::HeaderPt2 as ElfHeader;
-pub use xmas_elf::program::{ProgramHeader, ProgramIter, SegmentData};
+const ELF64_HDR_SIZE: usize = 64;
 
-#[derive(Debug)]
 pub struct ElfFile<'a> {
     elf_buf: &'a [u8],
-    elf_inner: xmas_elf::ElfFile<'a>,
+    elf_inner: Elf<'a>,
+    file_inode: Arc<dyn INode>,
+}
+
+impl<'a> Debug for ElfFile<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "ElfFile {{ inode: ???, elf_buf: {:?}, elf_inner: {:?} }}",
+            self.elf_buf, self.elf_inner,
+        )
+    }
 }
 
 impl<'a> ElfFile<'a> {
-    pub fn new(elf_buf: &'a [u8]) -> Result<ElfFile> {
-        let elf_inner =
-            xmas_elf::ElfFile::new(elf_buf).map_err(|e| errno!(ENOEXEC, "invalid ELF header"))?;
-        Self::validate(&elf_inner)?;
+    pub fn new(
+        file_inode: Arc<dyn INode>,
+        mut elf_buf: &'a mut [u8],
+        header: ElfHeader,
+    ) -> Result<ElfFile> {
+        let ctx = Ctx {
+            le: scroll::Endian::Little,
+            container: Container::Big,
+        };
+        let program_headers = ProgramHeader::parse(
+            elf_buf,
+            header.e_phoff as usize,
+            header.e_phnum as usize,
+            ctx,
+        )
+        .map_err(|e| errno!(ENOEXEC, "invalid program headers"))?;
 
-        Ok(ElfFile { elf_buf, elf_inner })
+        // read interpreter path
+        let mut count = 0;
+        let mut offset = 0;
+        let mut intepreter_count = 0;
+        let mut intepreter_offset = 0;
+        for ph in &program_headers {
+            ph.validate()?;
+            if ph.p_type == program_header::PT_INTERP && ph.p_filesz != 0 {
+                intepreter_count = (ph.p_filesz - 1) as usize;
+                intepreter_offset = ph.p_offset as usize;
+                println!(
+                    "PT_INTERP offset = {}, count = {}",
+                    intepreter_offset, intepreter_count
+                );
+                file_inode.read_at(
+                    intepreter_offset,
+                    &mut elf_buf[intepreter_offset..intepreter_offset + intepreter_count],
+                );
+                continue;
+            }
+
+            if ph.p_type == program_header::PT_LOAD {
+                count = (ph.p_filesz - 1) as usize;
+                offset = ph.p_offset as usize;
+                println!("PT_LOAD offset = {}, count = {}", offset, count);
+                file_inode.read_at(offset, &mut elf_buf[offset..offset + count]);
+            }
+        }
+
+        let interpreter = if intepreter_count == 0 {
+            None
+        } else {
+            str::from_utf8(&elf_buf[intepreter_offset..intepreter_offset + intepreter_count]).ok()
+        };
+        println!("interpreter = {:?}", interpreter);
+        let elf_inner =
+            goblin::elf::Elf::lazy_parse_for_occlum(header, program_headers, elf_buf, interpreter)
```

Comment:
Is there another lazy parser? If it is the only one, we may remove the word occlum in the name.

Reply:
Removed. Thank you for your review.